Control/Concurrent/MVar/Strict.hs

```diff
@@ -63,22 +66,27 @@ import Control.DeepSeq
 --     fairness properties of abstractions built using 'MVar's.
 --
 putMVar  :: NFData a => MVar a -> a -> IO ()
-#ifndef __HADDOCK__
-putMVar (MVar mvar#) !x = rnf x `seq` IO $ \ s# -> -- strict!
-    case putMVar# mvar# x s# of
-        s2# -> (# s2#, () #)
-#endif
+putMVar !mv !x = rnf x `seq` MV.putMVar mv x
```

Comment:
```suggestion
putMVar !mv x = rnf x `seq` MV.putMVar mv x
```
Gratuitous bang?

Reply:
That's needed to preserve the original behavior if `rnf` doesn't actually force anything.

```haskell
instance NFData Foo where
  rnf _ = ()
```

Reply:
What we *can* do is change that to `putMVar !mv x = force x `seq` MV.putMVar mv x`. I guess that's better.

Reply:
I'm afraid that subtlety is lost on me. `force x` is equivalent to ``rnf x `seq` x``, and that's materially different from `rnf x`? I'll take your word for it.

Reply:
Define

```haskell
data Foo
instance NFData Foo where
  rnf _ = ()
f, g :: Foo -> ()
f x = force x `seq` ()
g x = rnf x `seq` ()
```

Then `f undefined = undefined`, but `g undefined = ()`.

Whether this is the "right" decision is somewhat unclear to me, but it matches the previous behavior.

Reply:
Thanks for that example. The documentation for `rnf` says:
> rnf should reduce its argument to normal form (that is, fully evaluate all sub-components), and then return '()'.

so doesn't your example violate that? At any rate, this is indeed a pathological corner case. I certainly have no opinion.

Reply:
There are special circumstances when you might conceivably want `rnf` to do nothing, as a generic way to force the spine of a data structure but not its leaves. Do we want to worry about that?